src/ui/win/win_ui.cpp

```diff
@@ -0,0 +1,257 @@
+        throw imageException( "Unable to create Windows API window" );
+
+    RECT clientRoi;
+    GetClientRect( _window, &clientRoi );
+    RECT windowRoi;
+    GetWindowRect( _window, &windowRoi );
+
+    // Resize window to fit image 1 to 1
+    MoveWindow( _window, windowRoi.left, windowRoi.top,
+                windowRoi.right - windowRoi.left - (clientRoi.right - clientRoi.left) + image.width(),
+                windowRoi.bottom - windowRoi.top - (clientRoi.bottom - clientRoi.top) + image.height(), FALSE );
+}
+
+UiWindowWin::~UiWindowWin()
+{
+    _free();
+}
+
+void UiWindowWin::_display()
+{
+    if ( !_shown ) // we don't need to display anything
+        return;
+
+    ShowWindow( _window, SW_SHOWDEFAULT );
+    UpdateWindow( _window );
+
+    MSG windowMessage;
+    while ( GetMessage( &windowMessage, NULL, 0, 0 ) ) {
+        TranslateMessage( &windowMessage );
+        DispatchMessage( &windowMessage );
+    }
+}
+
+void UiWindowWin::setImage( const PenguinV_Image::Image & image )
+{
+    _free();
+
+    // We need to set image upside-down to show correctly in UI window as well as make 4 byte row length for bitmap
+    _image.setAlignment( 4u );
+    _image.setColorCount( image.colorCount() );
+    _image.resize( image.width(), image.height() );
+
+    const uint32_t rowSizeIn  = image.rowSize();
+    const uint32_t rowSizeOut = _image.rowSize();
+    const uint32_t width      = image.width() * image.colorCount();
+
+    const uint8_t * inY     = image.data() + image.rowSize() * (image.height() - 1);
+    uint8_t       * outY    = _image.data();
+    const uint8_t * outYEnd = outY + image.height() * rowSizeOut;
+
+    for ( ; outY != outYEnd; outY += rowSizeOut, inY -= rowSizeIn )
+        memcpy( outY, inY, sizeof( uint8_t ) * width );
+
+    const bool rgbImage = (image.colorCount() != 1u);
+    const DWORD bmpInfoSize = sizeof( BITMAPINFOHEADER ) + (rgbImage ? 1 : 256) * sizeof( RGBQUAD );
+
+    _bmpInfo = reinterpret_cast<BITMAPINFO*>(malloc( bmpInfoSize ));
+    _bmpInfo->bmiHeader.biSize          = sizeof( BITMAPINFOHEADER );
+    _bmpInfo->bmiHeader.biWidth         = image.width();
+    _bmpInfo->bmiHeader.biHeight        = image.height();
+    _bmpInfo->bmiHeader.biPlanes        = 1;
+    _bmpInfo->bmiHeader.biBitCount      = image.colorCount() * 8;
+    _bmpInfo->bmiHeader.biCompression   = BI_RGB;
+    _bmpInfo->bmiHeader.biSizeImage     = image.rowSize() * image.height();
+    _bmpInfo->bmiHeader.biXPelsPerMeter = 0;
+    _bmpInfo->bmiHeader.biYPelsPerMeter = 0;
+    _bmpInfo->bmiHeader.biClrUsed       = rgbImage ? 0 : 255;
+    _bmpInfo->bmiHeader.biClrImportant  = 0;
+
+    if ( !rgbImage ) {
+        for ( uint8_t i = 0u; ; ++i ) {
+            _bmpInfo->bmiColors[i].rgbRed   = i;
+            _bmpInfo->bmiColors[i].rgbGreen = i;
+            _bmpInfo->bmiColors[i].rgbBlue  = i;
+            _bmpInfo->bmiColors[i].rgbReserved = 0;
+
+            if ( i == 255u ) // to avoid variable overflow
```

Comment:
IMHO We should put the loop boundary in the `for` loop. `for ( size_t i = 0; i != 255 ; ++i )`

Reply:
It's done to avoid static cast for this part of code:
```cpp
            _bmpInfo->bmiColors[i].rgbRed   = i;
            _bmpInfo->bmiColors[i].rgbGreen = i;
            _bmpInfo->bmiColors[i].rgbBlue  = i;
```